### DISCUSSION TITLE
TS2345: Argument of type 'unknown' is not assignable to parameter of type 'DocumentInput '.

### DISCUSSION THREAD

**Discussion by SakisTsalk** [urql-graphql/urql#3148] [Q&A] [answered]
Hi, I've been trying to use React Urql with typescript by following the [documentation](https://formidable.com/open-source/urql/docs/basics/typescript-integration/).

I have generated the necessary schema code with codegen. And I'm trying to use this mutation in my component. 

``` 
const LoginMutation = graphql(/* GraphQL */`
  mutation login($loginCredentials: LoginUserInput!) {
    login(loginUserInput: $loginCredentials) {
      access_token
    }
  }
`);

  const [{ data }, login] = useMutation<Mutation>(LoginMutation);
```
But I'm getting a typescript error in useMutation: `TS2345: Argument of type 'unknown' is not assignable to parameter of type 'DocumentInput '.`

And that's because `graphql()` is of type unknown in the generated file.

```
/* eslint-disable */
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';

const documents = [];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 *
 *
 * @example
 * ```ts
 * const query = graphql(`query GetUser($id: ID!) { user(id: $id) { name } }`);
 * ```
 *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

export function graphql(source: string) {
  return (documents as any)[source] ?? {};
}

export type DocumentType<TDocumentNode extends DocumentNode<any, any>> =
  TDocumentNode extends DocumentNode<infer TType, any> ? TType : never;
```

What am I missing? Please update the documentation to include Mutation and subscription examples with typescript and codegen

**Comment by kitten** [comment]
Fwiw, it doesn't actually look like the `graphql` function is picking up your mutation definition, and the default of `unknown` is likely to make you aware of exactly this 🤔 I can't really help with this as I'm not too familiar with what may have changed in `@graphql-codegen/client-preset` and how it works.

We're currently working [on an alternative](https://github.com/0no-co/graphqlsp) that's more tailored for urql itself, but that's not ready yet and in an "alpha" stage.

You may want to check that the `documents` configuration includes the relevant files that the `graphql` function is called in, otherwise, I can only refer you to asking in the GraphQL Code Generator repo, unfortunately.

**Reply by SakisTsalk** [reply, marked as answer]
Hi Kitten, thanks for the quick reply.

I figured out my mistake by removing the `ignoreNoDocuments: true` flag that the documentation suggests for watch mode.

```
import { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  schema: '<YOUR_GRAPHQL_API_URL>',
  documents: ['src/**/*.ts'],
  ignoreNoDocuments: true, // for better experience with the watcher
  generates: {
    './src/gql/': {
      preset: 'client',
      plugins: [],
    },
  },
};

export default config;
```
My graphql mutation was in a `.ts` file so  it was not picked up by graphql codegen. After removing this flag the codegen script failed which pointed me to that mistake. Well, live and learn.
